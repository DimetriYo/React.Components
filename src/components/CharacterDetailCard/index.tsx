import { useNavigate, useParams } from 'react-router-dom';
import style from './style.module.css';
import { useEffect, useRef, useState } from 'react';
import { TCharacter } from '../../types';
import { fetchCharacterById } from '../../api/fetchCharacterById';
import { ContentLoader } from '../ContentLoader';

export function CharacterDetailCard() {
  const navigate = useNavigate();
  const { characterId } = useParams();
  const clickCounterRef = useRef(0);
  const [details, setDetails] = useState<TCharacter>();
  const [isLoading, setIsLoading] = useState(true);

  const hideDetails = () => {
    navigate({ pathname: '/', search: location.search });
  };

  const handleClickWhenCardShown = (e: MouseEvent) => {
    if (e.target === null) return;
    clickCounterRef.current++;
    const isOpenCardClick = !!(e.target as HTMLElement).closest('#character-details');
    const isNewCardClick = !!(e.target as HTMLElement).closest('[data-character-card]');
    if (!isOpenCardClick && !isNewCardClick && clickCounterRef.current > 1) hideDetails();
  };

  useEffect(() => {
    if (!characterId) return;
    fetchCharacterById(characterId)
      .then((character) => setDetails(character))
      .finally(() => setIsLoading(false));
    window.addEventListener('click', handleClickWhenCardShown);

    return () => window.removeEventListener('click', handleClickWhenCardShown);
  }, []);

  useEffect(() => {
    if (!characterId) return;
    fetchCharacterById(characterId)
      .then(setDetails)
      .finally(() => setIsLoading(false));
  }, [characterId]);

  if (!details) return;
  const { name, image, gender } = details;

  return (
    <>
      {isLoading && <ContentLoader />}
      <div data-testid="details-card" className={style.wrapper} id="character-details">
        <button
          onClick={hideDetails}
          className="absolute top-4 left-4 px-2 bg-blue-200 rounded-md border-2 border-solid border-slate-700 hover:bg-blue-600"
        >
          Close
        </button>
        <div className="flex flex-col gap-3 text-lg">
          <img
            src={image}
            className="rounded-3xl border-2 border-solid border-purple-600"
            alt={`${name} photo`}
          />
          <div>
            Name: <b>{name}</b>
          </div>
          <div>
            Gender: <b>{gender}</b>
          </div>
        </div>
      </div>
    </>
  );
}
