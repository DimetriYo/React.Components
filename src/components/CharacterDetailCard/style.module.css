.wrapper {
  @apply flex flex-col justify-center items-center relative basis-1/2 bg-orange-200 border-l-4 border-l-purple-700;
}
