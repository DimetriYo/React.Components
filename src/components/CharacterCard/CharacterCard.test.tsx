import { afterEach, describe, expect, it } from 'vitest';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { CharacterCard } from '.';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { CharacterDetailCard } from '../CharacterDetailCard';
import { testedCharacter } from '../../tests/testedCharacter';

afterEach(() => {
  cleanup();
});

describe('Character card', () => {
  it('should create card with mock data', () => {
    const { getByAltText } = render(
      <MemoryRouter>
        <CharacterCard {...testCharacterData} />
      </MemoryRouter>
    );

    expect(getByAltText('TestVasya photo')).toBeInTheDocument();
  });

  it('should open details on card click', async () => {
    const { getByRole, findByTestId, getByTestId } = render(
      <MemoryRouter>
        <Routes>
          <Route path="/" element={<CharacterCard {...testedCharacter} />} />
          <Route path="/:characterId" element={<CharacterDetailCard />} />
        </Routes>
      </MemoryRouter>
    );
    const card = getByRole('link');

    fireEvent.click(card);
    await findByTestId('details-card');

    expect(getByTestId('details-card')).toBeInTheDocument();
  });
});
