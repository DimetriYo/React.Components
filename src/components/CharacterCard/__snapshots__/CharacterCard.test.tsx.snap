// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Character card > should create card with mock data 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/12345"
      >
        <figure
          class="_characterCard_ce7192"
          data-character-card="true"
        >
          <img
            alt="TestVasya photo"
            src="/src/assets/nothingFound.png"
          />
          <figcaption
            class="text-center"
          >
            TestVasya
          </figcaption>
        </figure>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      href="/12345"
    >
      <figure
        class="_characterCard_ce7192"
        data-character-card="true"
      >
        <img
          alt="TestVasya photo"
          src="/src/assets/nothingFound.png"
        />
        <figcaption
          class="text-center"
        >
          TestVasya
        </figcaption>
      </figure>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
